src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/dependency/DependencyQueueTaskDispatcher.java

```diff
@@ -188,18 +188,40 @@ public CauseOfBlockage canRun(Queue.Item item) {
             return new BecauseDependentBuildIsBuilding(blockingProjects);
         } else {
             logger.info("No active dependencies on project: {} , it will now build", p);
+
+            ToGerritRunListener toGerritRunListener = ToGerritRunListener.getInstance();
```

Comment:
The above logging statement is a bit misleading, probably leading to this.
But returning `null` from a `QueueTaskDispatcher` does not necessarily mean that the build will run. It only means that **this** dispatcher thinks it's OK for the build to start. There could be other dispatchers later in the chain that will block it still.
And that means that this code can be called several times for an item in the queue and you'll end up adding multiple `GerritDependencyAction`s to the build.
Using either a `TransientActionFactory` or maybe a `QueueListener` would be better. You could also potentially add the one action to it when the trigger actually schedules the build and let it calculate itself when needed.

Reply:
Can I use `actions.replace` for now? I think the list of actions for Queue item should be relatively small (did not measure, btw). So, it will not be very expensive. In this case the rest of code will be the same.
Also it looks like in case of multiple actions we always pickup the first one, so, it means if we have the dependency A-B, we can have the race:
* A passed. This dispatcher added action.
* B was blocked because of other `QueueTaskDispatcher`
* A was re-triggered
* A passed, another dispatcher said ok
* Looks like we should see the first build of A - not a last one.

Reply:
`Actionable` has an [addOrReplace](http://javadoc.jenkins-ci.org/hudson/model/Actionable.html#addOrReplaceAction-hudson.model.Action-) that could work for now. But please add a TODO as well that this needs to be fixed.

Reply:
It's added only in `2.29`, but `replaceAction` called it inside anyway.